framework/src/main/java/org/checkerframework/common/value/util/Range.java

```diff
@@ -40,65 +42,81 @@
     public static boolean ignoreOverflow = false;
 
     /** A range containing all possible 64-bit values. */
-    public static final Range LONG_EVERYTHING = new Range(Long.MIN_VALUE, Long.MAX_VALUE);
+    public static final Range LONG_EVERYTHING = create(Long.MIN_VALUE, Long.MAX_VALUE);
 
     /** A range containing all possible 32-bit values. */
-    public static final Range INT_EVERYTHING = new Range(Integer.MIN_VALUE, Integer.MAX_VALUE);
+    public static final Range INT_EVERYTHING = create(Integer.MIN_VALUE, Integer.MAX_VALUE);
 
     /** A range containing all possible 16-bit values. */
-    public static final Range SHORT_EVERYTHING = new Range(Short.MIN_VALUE, Short.MAX_VALUE);
+    public static final Range SHORT_EVERYTHING = create(Short.MIN_VALUE, Short.MAX_VALUE);
 
     /** A range containing all possible char values. */
-    public static final Range CHAR_EVERYTHING = new Range(Character.MIN_VALUE, Character.MAX_VALUE);
+    public static final Range CHAR_EVERYTHING = create(Character.MIN_VALUE, Character.MAX_VALUE);
 
     /** A range containing all possible 8-bit values. */
-    public static final Range BYTE_EVERYTHING = new Range(Byte.MIN_VALUE, Byte.MAX_VALUE);
+    public static final Range BYTE_EVERYTHING = create(Byte.MIN_VALUE, Byte.MAX_VALUE);
 
-    /** The empty range. */
-    public static final Range NOTHING = new Range();
+    /** The empty range singleton. */
```

Comment:
In what way is this a singleton? I don't think we guarantee that this is the only Range object with that range.

Reply:
Two parts to this.

1. Actually this is from the original source code's line 98 (excerpt below) for the private constructor 
` /** Creates the singleton empty range. */`
` private Range() {`
` this.from = Long.MAX_VALUE;`
`this.to = Long.MIN_VALUE;`
	   
2. As written there is the guarantee  that no other existing or publicly created Range will have a different instance with these same bounds as the public factory method enforces from<=to which this special instance violates on purpose.

Reply:
Ah, I had missed that! Maybe this should be highlighted in the `Range()` constructor? It should only be called once and after that `NOTHING` should be used.

Reply:
@mernst handled that on the javadoc in this review.

It is getting better all the time. :)

Reply:
Now I looked at the new code again and you're doing exactly that. So this looks good now.

Reply:
It is better but I found an issue with the definition used for NOTHING : it does not allow a proper hash code to be computed.